<?xml version="1.0"?>
<launch>
  <arg name="world_name" default="$(find selective_grasping)/worlds/empty.world" doc="Default world"/>
  <arg name="paused" default="true" doc="Starts gazebo in paused mode" />
  <arg name="gui" default="true" doc="Starts gazebo gui" />

  ########################################
  # Load UR5 URDF
  ########################################
  <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find selective_grasping)/urdf/ur5_with_cam.urdf.xacro'"/>

  ########################################
  # Load Robot Description - GAZEBO
  ########################################

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(arg world_name)"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="paused" value="$(arg paused)"/>
    <arg name="headless" value="false"/>
    <arg name="gui" value="$(arg gui)"/>
  </include>

  <node pkg="selective_grasping" type="spawn_table.py" name="spawn_table_model"/>
  
  <node name="spawn_robot"
        pkg="gazebo_ros"
        type="spawn_model"
        args="-model robot
              -param robot_description
              -urdf
              -x 0
              -y 0
              -z 1.015
              -Y 1.57
              -unpause"
        respawn="false"
        output="screen" />  

  ########################################
  # Load Parameters
  ########################################
  <!-- Load joint names into the parameter server -->
  <rosparam file="$(find selective_grasping)/config/joint_names.yaml" command="load" />
  
  <!-- Launch GGCNN parameters -->
  <group ns="GGCNN">
    <rosparam file="$(find selective_grasping)/config/ggcnn_param.yaml" />
  </group>

  <!-- Parameters for object detection -->
  <rosparam file="$(find selective_grasping)/config/obj_detection.yaml" command="load" />

  ########################################
  # Load Gripper Controllers
  ########################################
  <!-- Load Robotiq Gripper Controller -->
  <rosparam file="$(find selective_grasping)/controller/ur5gripper_controllers.yaml"
            command="load"/>

  <node name="gripper_controller_spawner"
        pkg="controller_manager"
        type="spawner"
        args="gripper_controller_vel"/>

  
  ########################################
  # Load UR5 Controllers
  # load velocity and position controllers
  ########################################
  <rosparam file="$(find selective_grasping)/controller/ur5_controllers.yaml"
            command="load"/>

  <include file="$(find ur_gazebo)/launch/controller_utils.launch"/>

  <!-- LOAD CONTROLLERS AND RUN -->
  <arg name="controllers"
       default="pos_based_pos_traj_controller"/>

  <node name="ros_control_controller_spawner"
        pkg="controller_manager"
        type="spawner"
        respawn="false"
        output="screen"
        args="$(arg controllers)" />

  <!-- LOAD CONTROLLERS IN STOP MODE -->
  <arg name="stopped_controllers"
       default="joint_group_vel_controller gripper_controller_pos"/>

  <node name="ros_control_controller_manager"
        pkg="controller_manager"
        type="controller_manager"
        respawn="false"
        output="screen"
        args="load $(arg stopped_controllers)" />
</launch>